src/asynchronous/fw_update.rs

```diff
@@ -1,492 +1,692 @@
+    controllers: &mut [&mut T; N],
+) -> Result<[T::Guard; N], Error<T::BusError>> {
+    let mut guards = [const { None }; N];
 
-impl<'a> SliceImage<'a> {
-    pub fn new(image: &'a [u8]) -> Self {
-        Self { image, pos: 0 }
+    for (guard, controller) in zip(guards.iter_mut(), controllers.iter_mut()) {
+        *guard = Some(controller.disable_all_interrupts_guarded().await?);
     }
-}
 
-impl embedded_io_async::ErrorType for SliceImage<'_> {
-    type Error = embedded_io_async::ErrorKind;
+    Ok(guards.map(|g| g.unwrap()))
 }
 
-impl Read for SliceImage<'_> {
-    async fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
-        if buf.len() > self.image.len() - self.pos {
-            return Err(embedded_io_async::ErrorKind::InvalidInput);
-        }
+/// Re-enable interrupts on port 0 only, other ports don't respond to interrupts in FW update mode
+pub async fn enable_port0_interrupts<const N: usize, T: UpdateTarget>(
+    controllers: &mut [&mut T; N],
+) -> Result<[T::Guard; N], Error<T::BusError>> {
+    let mut guards = [const { None }; N];
 
-        buf.copy_from_slice(&self.image[self.pos..self.pos + buf.len()]);
-        self.pos += buf.len();
-
-        Ok(buf.len())
+    for (guard, controller) in zip(guards.iter_mut(), controllers.iter_mut()) {
+        *guard = Some(controller.enable_interrupt_guarded(PORT0, true).await?);
     }
+
+    Ok(guards.map(|g| g.unwrap()))
 }
 
-impl Seek for SliceImage<'_> {
-    async fn seek(&mut self, pos: SeekFrom) -> Result<u64, Self::Error> {
-        match pos {
-            SeekFrom::Start(p) => {
-                if p as usize > self.image.len() {
-                    return Err(embedded_io_async::ErrorKind::InvalidInput);
-                }
-                self.pos = p as usize;
-            }
-            SeekFrom::End(i) => {
-                if i > 0 || (-i) as usize > self.image.len() {
-                    return Err(embedded_io_async::ErrorKind::InvalidInput);
+/// General FW update function
+pub async fn perform_fw_update_borrowed<const N: usize, T: UpdateTarget>(
+    controllers: &mut [&mut T; N],
+    delay: &mut impl DelayNs,
+    pd_fw_bytes: &[u8],
+) -> Result<(), Error<T::BusError>> {
+    let mut updater = BorrowedUpdater::default();
+
+    // Disable all interrupts while we're entering FW update mode
+    let _guards = disable_all_interrupts(controllers).await?;
+    info!("Starting update");
+    let result = updater.start_fw_update(controllers, delay).await;
+    info!("Update started");
+
+    // Re-enable interrupts on port 0 only
+    let _guards = enable_port0_interrupts(controllers).await?;
+
+    match result {
+        Err(e) => {
+            error!("Failed to enter FW update mode");
+            Err(e)
+        }
+        Ok(mut updater) => {
+            info!("Sending chunks");
+            for chunk in pd_fw_bytes.chunks(UPDATE_CHUNK_LENGTH) {
+                if let Err(e) = updater.write_bytes(controllers, delay, chunk).await {
```

Comment:
`Ok(b)` isn't handled. Assuming I understand correctly, I think an `Ok(true)` should result in an early exit of the chunking loop.